File: Golang/Array/kadane's Algo/MaxCircular.go
package main

func max(a, b int)int{
	if a > b{
		return a
	}
	return b
}

func min(a, b int)int{
	if a < b{
		return a
	}
	return b
}

func maxCircular(arr[]int)int{
	totalSum, minSum, currMin, currMax := 0, 0, 0, 0
	maxSum := arr[0]

	for i := 0; i < len(arr) ; i++{
		currMax = max(currMax + arr[i], arr[i])
		maxSum = max(maxSum, currMax)

		currMin = min(currMin + arr[i], arr[i])
		minSum = min(currMin, minSum)

		totalSum += arr[i]
	}

	sum := maxSum
	circularSum := totalSum - minSum

	if minSum == totalSum{
		return sum
	}

	return max(sum, circularSum)
}

File: Java/Kadane_Algo/MaxCircular.java
/**You are given a circular array arr[] of integers, find the maximum possible sum of a non-empty subarray. 
 * In a circular array, the subarray can start at the end and wrap around to the beginning. Return the maximum 
 * non-empty subarray sum, considering both non-wrapping and wrapping cases. */
public class MaxCircular{
    public static int maxCircularSum(int[] arr){
        int totalSum = 0;
        int maxSum = arr[0], minSum = 0;
        int currMax = 0, currMin = 0;

        for(int i = 0; i < arr.length; i++){
            currMax = Math.max(currMax + arr[i], arr[i]);
            maxSum = Math.max(maxSum, currMax);

            currMin = Math.min(currMin + arr[i], arr[i]);
            minSum = Math.min(minSum, currMin);

            totalSum += arr[i];
        }

        int sum = maxSum;
        int circularSum = totalSum - minSum;

        if(minSum == totalSum) return sum;

        return Math.max(sum, circularSum);
    }
    public static void main(String[] args) {
       int[] nums = {8, -8, 9, -9, 10, -11, 12};
       System.out.println(MaxCircular.maxCircularSum(nums));
    }
}